[OSDOCS#3671]: Gathering and interpreting cv history (#69380)
openshift/openshift-docs: modules/gathering-clusterversion-history.adoc

// Module included in the following assemblies:
//
// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc

:_mod-docs-content-type: PROCEDURE
[id="gathering-clusterversion-history_{context}"]
= Gathering ClusterVersion history 

> The cluster version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 

Reviewer: ```suggestion
The Cluster Version Operator (CVO) records updates made to a cluster, known as the ClusterVersion history. The entries can reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
```

Reviewer: Capitalize Operator names. See https://redhat-documentation.github.io/supplementary-style-guide/#operator

[NOTE]
====

> The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit. 

Reviewer 2: hi @obrown1205 Shall we put more details about the limitation in the doc? like below in related bug, at least put the limit size 100 here:
```
Initial version is kept
all minor version updates are kept
the latest z streams updates are kept
however, the oldest z stream updates (in previous minor versions), which are not considered as important to the overall history of the cluster, are deleted from the history
```

Author: can do :)

Author: @LalatenduMohanty what do you think of this?

The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the oldest z stream updates in previous minor versions are pruned to accommodate the limit.

Reviewer 3: This looks mostly correct . But I do not think we need to go for writing the details. Anyone who will go through history should know that we might have modified the history if the number of entries are 100.

> The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the oldest z stream updates in previous minor versions are pruned to accommodate the limit. 

Reviewer: suggestion:

```suggestion
The initial, minor, and z-stream version updates are stored by the ClusterVersion history. However, the ClusterVersion history has a size limit. If the limit is reached, the oldest z-stream updates in previous minor versions are pruned to accommodate the limit. 
```

====

> It is possible to view the ClusterVersion history by using the {product-title} web console and using the {oc-first}.

Reviewer: ```suggestion
You can view the ClusterVersion history by using the {product-title} web console or by using the {oc-first}.
```

// Module included in the following assemblies:
//
// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc

:_mod-docs-content-type: PROCEDURE
[id="gathering-clusterversion-history_{context}"]
= Gathering ClusterVersion history 

> When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update. To gather and understand the history of your successful or failed cluster update, it is important to retrieve the ClusterVersion history.

Reviewer 4: nit: "failed" sounds so final.  The CVO has `Failing`, which is "we're currently having trouble", but the CVO will never give up on trying to reconcile the target release it's been pointed at.  And the `To gather...` sentence here seems a bit tautological ("to gather the history, you need to gather the history"), without explaining how gathering the history will help accomplish something else.  Maybe rephrase to something like:

> When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update.  This can be useful to decide if history-related bugs apply to your cluster.  It can also be used to correlate changes in cluster behavior with one potential trigger, although correlation alone does not imply causation.

[NOTE]
====
The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit. 
====

.Prerequisites

* You have access to the cluster as a user with the `cluster-admin` role.

> * You have installed the OpenShift CLI (`oc`).

Reviewer 5: We can also mention that web console shows this in its "Administration -> Cluster Settings" section

.Procedure

> . View a detailed summary of cluster specifications, update availability, and update history:

Reviewer 4: nit: looking at `.status.history` just gives you "update history".  It would be `availableUpdates` for availability, and... something under `spec` for cluster specifications?  "detailed summary" also seems like it's pulling in two directions, and `history` seems more on the detailed side of things and less on the summary side of things.  Maybe something like:

> View cluster update history:

?

+
[source,terminal]
----

> $ oc get clusterversion version -o json | jq .status.history

Reviewer 4: Dropping the `jq` exposure (you don't list it as a prereq, and reducing the number of prereqs is good) with [JSONPath][1]:

```console
$ oc get -o 'jsonpath={range .status.history[*]}{@}{"\n"}{end}' clusterversion version
{"completionTime":"2023-11-07T20:26:04Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4","startedTime":"2023-11-07T19:11:36Z","state":"Completed","verified":true,"version":"4.13.19"}
{"completionTime":"2023-10-04T18:53:29Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec","startedTime":"2023-10-04T17:26:31Z","state":"Completed","verified":true,"version":"4.13.13"}
{"completionTime":"2023-09-26T14:21:43Z","image":"quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727","startedTime":"2023-09-26T14:02:33Z","state":"Completed","verified":false,"version":"4.13.12"
```

That's not quite as pretty as the newlined output that `jq` produces, but it's not completely illegible.  Another alternative with more opinions about how to format the properties in each entry:

```console
$ oc get -o 'jsonpath={range .status.history[*]}{.startedTime}{" "}{.completionTime}{" "}{.state}{" "}{.version}{" (verified: "}{.verified}{", "}{.image}{") "}{.acceptedRisks}{"\n"}{end}' clusterversion version
2023-11-07T19:11:36Z 2023-11-07T20:26:04Z Completed 4.13.19 (verified: true, quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4) 
2023-10-04T17:26:31Z 2023-10-04T18:53:29Z Completed 4.13.13 (verified: true, quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec) 
2023-09-26T14:02:33Z 2023-09-26T14:21:43Z Completed 4.13.12 (verified: false, quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727)
```

There's also:

```console
$ oc describe clusterversions/version
...
  Desired:
    Channels:
      candidate-4.13
      candidate-4.14
      fast-4.13
      fast-4.14
      stable-4.13
    Image:    quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4
    URL:      https://access.redhat.com/errata/RHSA-2023:6130
    Version:  4.13.19
  History:
    Completion Time:    2023-11-07T20:26:04Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:a148b19231e4634196717c3597001b7d0af91bf3a887c03c444f59d9582864f4
    Started Time:       2023-11-07T19:11:36Z
    State:              Completed
    Verified:           true
    Version:            4.13.19
    Completion Time:    2023-10-04T18:53:29Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:eac141144d2ecd6cf27d24efe9209358ba516da22becc5f0abc199d25a9cfcec
    Started Time:       2023-10-04T17:26:31Z
    State:              Completed
    Verified:           true
    Version:            4.13.13
    Completion Time:    2023-09-26T14:21:43Z
    Image:              quay.io/openshift-release-dev/ocp-release@sha256:371328736411972e9640a9b24a07be0af16880863e1c1ab8b013f9984b4ef727
    Started Time:       2023-09-26T14:02:33Z
    State:              Completed
    Verified:           false
    Version:            4.13.12
  Observed Generation:  4
  Version Hash:         CMLl3sLq-EA=
Events:                 <none>
```

if we think that pretty-formatting of the `history` part is worth the hassle of scrolling through the non-history part of the structure.

And if we expect this to be something that folks need with regularity, in isolation from other parts of ClusterVersion, we can add new `oc ...` subcommands to make it easy to get at this information without needing `jq` or JSONPath.

[1]: https://kubernetes.io/docs/reference/kubectl/jsonpath/

Reviewer 4: Also, these options are not exclusive.  You could present both a JSONPath example and a `jq` example if you like.

// Module included in the following assemblies:
//
// * updating/troubleshooting_updates/gathering-data-cluster-update.adoc

:_mod-docs-content-type: PROCEDURE
[id="gathering-clusterversion-history_{context}"]
= Gathering ClusterVersion history 

> When updating your cluster, the cluster version operator (CVO) creates a `status.history` entry to record the update. This can be useful to determine if history-related bugs apply to your cluster. The entry can also reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 

Reviewer 6: ```suggestion
When updating your cluster, the cluster version Operator (CVO) creates a `status.history` entry to record the update. This can be useful to determine if history-related bugs apply to your cluster. The entry can also reveal correlation between changes in cluster behavior with potential triggers, although correlation does not imply causation. 
```
"Operator" should be capitalized - I also wonder, should it be "Cluster Version Operator" if that's an official name?

Suggestion: It may be worth adding an attribute for the Operator name if one doesn't already exist.

Reviewer 6: Would it be helpful to users to mention here where the "`status.history` entry" is created? Is this part of some system logs, etc?

Author: good call - I'll confirm with the engineers :)

Author: I've removed that sentence since it was more confusing than accurate. :) thanks!

[NOTE]
====
The ClusterVersion history has a size limit, so in cases where enough updates occurred to have reached that limit, the least relevant entries in the history are pruned to accommodate the limit. 
====

> It is possible to view the ClusterVersion history though the web console and through the CLI.

Reviewer 6: ```suggestion
It is possible to view the ClusterVersion history in the {product-title} web console, or by using the {oc-first}.
```
Per the IBMSG, do not use "through" to mean "by using":
![image](https://github.com/openshift/openshift-docs/assets/40172997/8f075877-6a24-4ab2-874f-22d9f0d4ee3f)

https://www.ibm.com/docs/en/ibm-style?topic=word-usage#through

Reviewer 6: Also suggest using the attributes here for the web console and CLI names

To view the ClusterVersion history through the web console:

.Prerequisites
* You have access to the cluster as a user with the `cluster-admin` role.
* You have access to the {product-title} web console.

.Procedure

. From the web console, click *Administration* -> *Cluster Settings* and review the contents of the *Details* tab.

> To view the ClusterVersion history through the CLI:

Reviewer 6: ```suggestion
To view the ClusterVersion history through the {oc-first}:
```

.Prerequisites
* You have access to the cluster as a user with the `cluster-admin` role.

> * You have installed the OpenShift CLI (`oc`).

Reviewer 6: ```suggestion
* You have installed the {oc-first}.
```

.Procedure

> 

Reviewer 6: There should only be one list of Prerequisites and one Procedure per module. This should probably be split into 2 modules, or reformatted so that the steps are maybe optional steps in a single procedure.

> . View cluster update history:

Reviewer 6: If it's just one step in a procedure, use a bullet instead of a numbered list: 
![image](https://github.com/openshift/openshift-docs/assets/40172997/820de4dd-91e4-42d2-b804-258249471cf6)

https://redhat-documentation.github.io/modular-docs/#con-creating-procedure-modules_writing-mod-docs

+
[source,terminal]
----
$ oc describe clusterversions/version
----
+
.Example output
[source,terminal]
----
  $ oc describe clusterversions/version

> ...

Reviewer 6: Do not include the command in the example output code block:

![image](https://github.com/openshift/openshift-docs/assets/40172997/74798f97-91b7-49a1-ac24-045ef080cd2a)

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output
